### PR TITLE
[Feature](count_by_enum) support count_by_enum function

### PR DESCRIPTION
**Pull request by m3ngzhang**
## Proposed changes

Issue Number: close #xxx

## Problem summary
`count_by_enum(expr1, expr2, ... , exprN);`

Treats the data in a column as an enumeration and counts the number of values in each enumeration. Returns the number of enumerated values for each column, and the number of non-null values versus the number of null values.

#### Arguments

`expr1` — At least one input must be specified. The value is a column of type `STRING`.

##### Returned value

Returns a JSONArray string.

For example:
```json
[{
  "cbe": {
    "F": 100,
    "M": 99
  },
  "notnull": 199,
  "null": 1,
  "all": 200
}, {
  "cbe": {
    "20": 10,
    "30": 5,
    "35": 1
  },
  "notnull": 16,
  "null": 184,
  "all": 200
}, {
  "cbe": {
    "China": 10,
    "United States": 9,
    "England": 20,
    "Germany": 30
  },
  "notnull": 69,
  "null": 131,
  "all": 200
}]
```
Description: The return value is a JSON array string and the order of the internal objects is the order of the input parameters.
* cbe: count of non-null values based on enumeration values
* notnull: number of non-null values.
* null: number of null values
* all: total number, including both null and non-null values.

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by czzmmc** [comment]
@zhangstar333 I falsely closed the PR(#21818). So I recreated this one, please check.

File: be/src/vec/aggregate_functions/aggregate_function_count_by_enum.h [hunk 1]
```diff
@@ -0,0 +1,202 @@
...
+
+    std::string get_name() const override { return "count_by_enum"; }
+
+    DataTypePtr get_return_type() const override { return std::make_shared<DataTypeString>(); }
+
+    void add(AggregateDataPtr __restrict place, const IColumn** columns, size_t row_num,
+             Arena* arena) const override {
+
+        for (int i = 0; i < arg_count; i++) {
+            const auto* nullable_column = check_and_get_column<ColumnNullable>(columns[i]);
```
[the remark this answers is not in the document]
Comment (czzmmc) [reply be/src/vec/aggregate_functions/aggregate_function_count_by_enum.h:166]: We wanna count all rows including not null rows and null rows.

File: be/src/vec/aggregate_functions/aggregate_function_count_by_enum.h [hunk 2]
```diff
@@ -0,0 +1,202 @@
...
+#include "vec/utils/count_by_enum_helpers.hpp"
+
+namespace doris::vectorized {
+
+struct AggregateFunctionCountByEnumData {
+    using MapType = std::unordered_map<std::string, uint64_t>;
+
+    void reset() { data_vec.clear(); }
+
+    void add(int idx, const StringRef& value, const UInt64& number = 1) {
```
[the remark this answers is not in the document]
Comment (czzmmc) [reply be/src/vec/aggregate_functions/aggregate_function_count_by_enum.h:40]: In our condition, yes.

File: be/src/vec/aggregate_functions/aggregate_function_count_by_enum.h (hunk 1 above)
[the remark this answers is not in the document]
Comment (zhangstar333) [reply be/src/vec/aggregate_functions/aggregate_function_count_by_enum.h:166]: const auto* nullable_column = check_and_get_column<ColumnNullable>(columns[i]);
if columns[I] is not nullable column, the nullable_column = nullptr

[the remark this answers is not in the document]
Comment (czzmmc) [reply be/src/vec/aggregate_functions/aggregate_function_count_by_enum.h:166]: I fix this problem, please check.

File: be/src/vec/utils/count_by_enum_helpers.hpp [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+        obj.AddMember("null", data_vec[idx].null, allocator);
+        obj.AddMember("all", data_vec[idx].all, allocator);
+
+        doc.PushBack(obj, allocator);
+    }
+
+    rapidjson::Writer<rapidjson::StringBuffer> writer(buffer);
+    doc.Accept(writer);
+
+    return buffer.GetSize() > 0;
```
[the remark this answers is not in the document]
Comment (czzmmc) [reply be/src/vec/utils/count_by_enum_helpers.hpp]: Fixed in the latest commit.

File: be/src/vec/utils/count_by_enum_helpers.hpp [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+
+    int vec_size_number = data_vec.size();
+    for (int idx = 0; idx < vec_size_number; ++idx) {
+        rapidjson::Value obj(rapidjson::kObjectType);
+
+        rapidjson::Value obj_cbe(rapidjson::kObjectType);
+        std::unordered_map<std::string, uint64_t> unordered_map = data_vec[idx].cbe;
+        for (auto it : unordered_map) {
+            std::stringstream ss;
+            value_to_string(ss, it.first);
```
[the remark this answers is not in the document]
Comment (czzmmc) [reply be/src/vec/utils/count_by_enum_helpers.hpp]: Fixed in the latest commit.

**Comment by zhangstar333** [comment]
it's ok for BE part, @morrySnow PTAL about nereids. thanks

and I thinks this feature could push for master firstly, then pick to 1.2lts branch.

File: be/src/vec/aggregate_functions/aggregate_function_count_by_enum.cpp [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+    }
+
+    LOG(WARNING) << fmt::format("unsupported input type {} for aggregate function {}",
+                                argument_types[0]->get_name(), name);
+    return nullptr;
+}
+
+void register_aggregate_function_count_by_enum(AggregateFunctionSimpleFactory& factory) {
+    factory.register_function("count_by_enum", create_aggregate_function_count_by_enum, true);
+}
```
Comment (zhangstar333) [inline be/src/vec/aggregate_functions/aggregate_function_count_by_enum.cpp:58]: here maybe need register for not nullable also.
`factory.register_function("count_by_enum", create_aggregate_function_count_by_enum, false);`

**Comment by czzmmc** [comment]
> it's ok for BE part, @morrySnow PTAL about nereids. thanks
> 
> and I thinks this feature could push for master firstly, then pick to 1.2lts branch.

@zhangstar333 I created a PR (https://github.com/apache/doris/pull/22071) for master.

**Pull request closed without merging**